[PT][FSDP] support custom all reduce hook across FSDP units

xunnanxu opened the pull request:
This change adds an API `set_all_reduce_hook` to the `FSDPModule` to
support customized all reduce either in native HSDP (2d mesh) setup or custom HSDP (1d FSDP + custom AR across replicas)
* For native HSDP, the original AR would still run as is and this hook allows for additional gradient modification post all reduce.
* For custom HSDP, the original AR will be skipped and all the logic is instead expected to be executed in the hook.

The custom hook is expected to perform operations in place (no return value).

Example basic usage:
```
model = ...
fully_shard(model, mesh=...)
model.set_all_reduce_hook(my_hook)
```

By default, the hook will run in the default all reduce stream post reduce scatter.
When native HSDP is NOT enabled, the custom hook can be specified to run in a custom stream. This custom stream will also be synchronized post reduce scatter similarly. See tests for examples.

Test Plan: CI

Differential Revision: D68255583




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

awgu commented on torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:
```diff
@@ -437,6 +438,18 @@ def foreach_reduce(
+    # -- END: ops in reduce_scatter stream
+
+    if all_reduce_hook is not None:
+        # Execute user-specified all reduce hook across FSDP units.
```
super nit: the `across FSDP units` part is a bit confusing to me since the 'across' could mean both "same FSDP unit but across DP ranks" or "same DP rank but across FSDP units on this rank"

maybe can just exclude the "across FSDP units" part -- it should be clear the all-reduce will be across some DP ranks

xunnanxu replied:
Understood. Naming is hard :(
Updated.

xunnanxu commented:
@pytorchbot merge

xunnanxu commented:
@pytorchmergebot merge -r main

xunnanxu commented:
@pytorchbot merge -r main

xunnanxu commented:
@pytorchbot merge

xunnanxu committed:
[PT][FSDP] support custom all reduce hook across FSDP units (#147114)

Summary:
This change adds an API `set_all_reduce_hook` to the `FSDPModule` to
support customized all reduce either in native HSDP (2d mesh) setup or custom HSDP (1d FSDP + custom AR across replicas)
* For native HSDP, the original AR would still run as is and this hook allows for additional gradient modification post all reduce.
* For custom HSDP, the original AR will be skipped and all the logic is instead expected to be executed in the hook.

The custom hook is expected to perform operations in place (no return value).

Example basic usage:
```
model = ...
fully_shard(model, mesh=...)
model.set_all_reduce_hook(my_hook)
```

By default, the hook will run in the default all reduce stream post reduce scatter.
When native HSDP is NOT enabled, the custom hook can be specified to run in a custom stream. This custom stream will also be synchronized post reduce scatter similarly. See tests for examples.


Test Plan:
CI





cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Reviewed By: awgu, ckluk2

Differential Revision: D68255583

Pulled By: xunnanxu
